rfd/0131-adminitrative-actions-mfa.md

```diff
@@ -0,0 +1,532 @@
+- `SetNetworkRestrictions`, `DeleteNetworkRestrictions`
+- `UpsertLock`, `DeleteLock`, `ReplaceRemoteLocks`
+- `CreateApp`, `UpdateApp`, `DeleteApp`, `DeleteAllApps`
+- `CreateDatabase`, `UpdateDatabase`, `DeleteDatabase`, `DeleteAllDatabases`
+- `UpsertWindowsDesktopService`, `DeleteWindowsDesktopService`, `DeleteAllWindowsDesktopServices`
+- `CreateWindowsDesktop`, `UpdateWindowsDesktop`, `UpsertWindowsDesktop`, `DeleteWindowsDesktop`, `DeleteAllWindowsDesktops`
+- `ChangeUserAuthentication`, `StartAccountRecovery`, `VerifyAccountRecovery`, `CompleteAccountRecovery`, `CreateAccountRecoveryCodes`
+- `CreatePrivilegeToken`, `CreateRegisterChallenge`
+- `GenerateCertAuthorityCRL`
+- `CreateConnectionDiagnostic`, `UpdateConnectionDiagnostic`, `AppendDiagnosticTrace`
+- `SetInstaller`, `DeleteInstaller`, `DeleteAllInstallers`
+- `SetUIConfig`, `DeleteUIConfig`
+- `CreateKubernetesCluster`, `UpdateKubernetesCluster`, `DeleteKubernetesCluster`, `DeleteAllKubernetesClusters`
+- `CreateSAMLIdPServiceProvider`, `UpdateSAMLIdPServiceProvider`, `DeleteSAMLIdPServiceProvider`, `DeleteAllSAMLIdPServiceProviders`
+- `CreateUserGroup`, `UpdateUserGroup`, `DeleteUserGroup`, `DeleteAllUserGroups`
+- `UpdateHeadlessAuthenticationState`
+- `ExportUpgradeWindows`
+- `UpdateClusterMaintenanceConfig`
+- `UpdatePluginData`
+- `CreateDevice`, `UpdateDevice`, `UpsertDevice`, `DeleteDevice`, `BulkCreateDevices`, `CreateDeviceEnrollToken`
+
+HTTP:
+
+- `rotateCertAuthority`, `rotateExternalCertAuthority`
+- `generateHostCert`
+- `createWebSession`, `deleteWebSession`
+- `upsertAuthServer`, `upsertProxy`
+- `deleteAllProxies`, `deleteProxy`
+- `upsertTunnelConnection`, `deleteTunnelConnection`, `deleteTunnelConnections`, `deleteAllTunnelConnections`
+- `createRemoteCluster`, `deleteRemoteCluster`, `deleteAllRemoteClusters`
+- `upsertReverseTunnel`, `deleteReverseTunnel`
+- `validateTrustedCluster`, `registerUsingToken`
+- `deleteNamespace`
+- `setClusterName`
+- `deleteStaticTokens`, `setStaticTokens`
+- `validateGithubAuthCallback`
+- `upsertCertAuthority`, `deleteCertAuthority`
+- `deleteUser`
+
+Notable exceptions:
+
+- User actions that are authorized based on the user owning the resource.
+These requests will only require MFA when the authorization comes from the
+user's role rather than the user itself. For example:
+  - `CreateAccessRequest`, `DeleteAccessRequest`
+  - `CreateAuthenticateChallenge`
+  - `ChangePassword`, `CreateResetPasswordToken`
+  - `AddMFADeviceSync`, `DeleteMFADeviceSync`
+- Actions which only require `DefaultImplicitRole`:
+  - `SubmitUsageEvent`
+- Actions which are limited to internal service roles. For example:
+  - `CreateSessionTracker`, `RemoveSessionTracker`, `UpdateSessionTracker`
+  - `GenerateWindowsDesktopCert`, `GenerateOpenSSHCert`
+  - `InventoryControlStream`
+  - `SendKeepAlives`
+  - `EmitAuditEvent`, `CreateAuditStream`, `ResumeAuditStream`
+  - `ProcessKubeCSR`, `SignDatabaseCSR`
```

Comment:
Users can call SignDatabaseCSR as well when doing `tctl auth sign` if I remember correctly. Why do these have to be exceptions, if we don't require MFA for builtin roles (RoleAdmin, RoleDatabase, etc.) then these API calls would automatically not require it for agent calls and will require it for user calls?

Reply:
This isn't an exhaustive list, but these are endpoints which are only usable through built-in roles. This is different from endpoints which require specific allow rules _or_ a built in role. Also, from what I can tell, `SignDatabaseCSR` can only be called by the Proxy.